### PR TITLE
tidb-lightning: add 50 TB data import best practices

### PR DESCRIPTION
**Pull request by niubell**
### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v7.2 (TiDB 7.2 versions)
- [ ] v7.1 (TiDB 7.1 versions)
- [ ] v7.0 (TiDB 7.0 versions)
- [ ] v6.6 (TiDB 6.6 versions)
- [ ] v6.5 (TiDB 6.5 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: tidb-lightning/tidb-lightning-50TB-import-best-practices.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
+---
+title: 50 TB 数据导入最佳实践
+summary: 本文根据大单表导入的经验，总结出了一套导入大量数据的最佳实践，希望对大数据量、大单表导入有所帮助。
```
Comment (Frank945946) [inline tidb-lightning/tidb-lightning-50TB-import-best-practices.md:3]:
```suggestion
summary: 本文根据导入大单表的经验，总结出了一套导入大量数据的最佳实践，希望对你在导入大数据量、大单表的场景有所帮助。
```

File: tidb-lightning/tidb-lightning-50TB-import-best-practices.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+---
+
+# 50 TB 数据导入最佳实践
+
+TiDB Lightning（[物理导入模式](/tidb-lightning/tidb-lightning-physical-import-mode.md)）是一款用于空表导入、空集群初始化的全量和高效的数据导入工具。并且 TiDB Lightning 以文件作为数据源。TiDB Lightning 提供了两种运行方式：单实例和[并行导入](/tidb-lightning/tidb-lightning-distributed-import.md)，满足不同规模的源文件导入。
```
Comment (Frank945946) [inline tidb-lightning/tidb-lightning-50TB-import-best-practices.md:8]:
```suggestion
TiDB Lightning（[物理导入模式](/tidb-lightning/tidb-lightning-physical-import-mode.md)）是一款用于将离线数据导入空表、空集群的高效的数据导入工具。并且 TiDB Lightning 以文件作为数据源。TiDB Lightning 提供了两种运行方式：单实例和[并行导入](/tidb-lightning/tidb-lightning-distributed-import.md)，满足不同规模的源文件导入。
```

File: tidb-lightning/tidb-lightning-50TB-import-best-practices.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
+---
+title: 50 TB 数据导入最佳实践
```
Comment (Frank945946) [inline tidb-lightning/tidb-lightning-50TB-import-best-practices.md:2]: 全文单位都统一成 TiB 吧。

File: tidb-lightning/tidb-lightning-50TB-import-best-practices.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+- [导入大单表的最佳实践](#导入大单表的最佳实践)
+
+## 关键因素
+
+在导入数据时，有一些关键因素会影响导入性能，甚至可能导致导入失败。一些常见的关键因素如下：
+
+- 源文件
+
+    - 单个文件内数据是否按照主键有序。有序可以达到最优导入性能。
+    - 多个 TiDB Lightning 实例对应的源文件内容是否按照主键有交叠。交叠越小，导入性能越好。
```
Comment (Frank945946) [inline tidb-lightning/tidb-lightning-50TB-import-best-practices.md:37]:
```suggestion
    - 多个 TiDB Lightning 实例导入的源文件之间，是否存在重叠的主键或者非空唯一索引。重叠越少，导入性能越好。
```

File: tidb-lightning/tidb-lightning-50TB-import-best-practices.md [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+    - [Issue-14745](https://github.com/tikv/tikv/issues/14745)：导入完成后 TiKV Import 目录遗留大量临时文件。
+    - [Issue-6426](https://github.com/tikv/pd/issues/6426)：PD [范围调度](/tidb-lightning/tidb-lightning-physical-import-mode-usage.md#导入时暂停-pd-调度的范围)接口存在未打散 Region 的情况，导致 Scatter Region 超时。v6.2.0 之前采用停止全局调度的方式，不会出现该问题。
+    - [Issue-43079](https://github.com/pingcap/tidb/pull/43079)：TiDB Lightning 对 NotLeader 错误的重试未刷新 Region Peers 信息。
+    - [Issue-43291](https://github.com/pingcap/tidb/issues/43291)：TiDB Lightning 未对临时文件删除的情况（"No such file or directory"）进行重试。
+
+## 准备源文件
+
+- 生成文件时，单个文件内，尽量按照主键排序；如果表定义没有主键，可以添加一个自增主键，此时对文件内容顺序无要求。
+- 多个 TiDB Lightning 实例在划分源文件时，尽量降低主键交叠。如果生成文件时全局有序，可以按照范围划分文件给不同 TiDB Lightning 实例，达到最佳导入效果。
```
Comment (Frank945946) [inline tidb-lightning/tidb-lightning-50TB-import-best-practices.md:79]:
```suggestion
- 给多个 TiDB Lightning 实例分配要导入的源文件时，尽量避免多个源文件之间存在重叠的主键或非空唯一索引的情况。如果生成文件是全局有序，可以按照范围划分不同的文件给不同 TiDB Lightning 实例进行导入，达到最佳导入效果。
```

File: tidb-lightning/tidb-lightning-50TB-import-best-practices.md [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+如果导入过程中遇到下游 TiKV 空间不足，可以手动执行 `kill` 命令关闭（不要带 `-9` 选项）所有 TiDB Lightning 实例，待扩容后，基于断点信息继续导入。
+
+## 导入大单表的最佳实践
+
+多表导入会带来 Checksum、Analyze 时间的增加，甚至超过数据导入本身，但是一般不需要调整配置。如果多表中存在单个或多个大表的情况，可以把这类大表的源文件划分出来，单独进行导入。
+
+本小节重点介绍大单表导入的最佳实践。大单表没有严格的定义，一般认为符合以下任一条件者即为大单表：
+
+- 大小超过 10 TB
+- 行数超过 10 亿、列超过 50 的宽表
```
Comment (Frank945946) [inline tidb-lightning/tidb-lightning-50TB-import-best-practices.md:124]:
```suggestion
- 行数超过 10 亿、列数超过 50 的宽表
```

File: tidb-lightning/tidb-lightning-50TB-import-best-practices.md [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+### 准备源文件
+
+根据上述源文件准备的步骤产生源文件，对于大单表，如果你不能做到全局有序，但是可以做到文件内按主键有序，且是标准的 CSV 文件，可以尽量生成单个大文件（每个 20 GB），然后开启 [strict-format](/migrate-from-csv-files-to-tidb.md#第-4-步导入性能优化可选)，既可以降低 TiDB Lightning 实例之间的交叠，又能在导入前由 TiDB Lightning 进行大文件切分，达到最佳的导入速度。
```
Comment (Frank945946) [inline tidb-lightning/tidb-lightning-50TB-import-best-practices.md:128]:
```suggestion
根据上述源文件准备的步骤产生源文件，对于大单表，如果你不能做到全局有序，但是可以做到文件内按主键有序，且是标准的 CSV 文件，可以尽量生成单个大文件（每个 20 GB），然后开启 [strict-format](/migrate-from-csv-files-to-tidb.md#第-4-步导入性能优化可选)，既可以降低 TiDB Lightning 实例之间导入的数据文件中存在主键和唯一键的重叠，又能在导入前由 TiDB Lightning 实例对大文件进行切分，达到最佳的导入速度。
```

File: tidb-lightning/tidb-lightning-50TB-import-best-practices.md (hunk 3 above)
Comment (hfxsd) [reply tidb-lightning/tidb-lightning-50TB-import-best-practices.md]: done

File: tidb-lightning/tidb-lightning-50TB-import-best-practices.md [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+- 预先导入 10% 的数据，实际占用空间再乘以 10，即可认为是该批数据最终的空间占用。该方法更加准确，尤其是对于导入大量数据时比较有效。
+
+注意要预留 20% 的存储空间，后台任务如压缩、复制快照等会使用部分存储空间。
+
+## 配置参数
+
+需要正确设置以下配置参数：
+
+- `region-concurrency`：TiDB Lightning 主逻辑处理的并发度。在并行导入时，可以设置为 CPU 核数的 75%，防止出现资源过载带来 OOM 问题。
+- `send-kv-pairs`：TiDB Lightning 发送给 TiKV 单次请求的 Key、Value 数量，建议按照 send-kv-pairs * row-size < 1 MB 调整该值。v7.2.0 版本会用 `send-kv-size` 代替该参数，且无需单独设置。
```
Comment (hfxsd) [inline tidb-lightning/tidb-lightning-50TB-import-best-practices.md:97]:
```suggestion
- `send-kv-pairs`：TiDB Lightning 发送给 TiKV 单次请求的 Key、Value 数量，建议按照 send-kv-pairs * row-size < 1 MiB 调整该值。v7.2.0 版本会用 `send-kv-size` 代替该参数，且无需单独设置。
```

File: tidb-lightning/tidb-lightning-50TB-import-best-practices.md [hunk 9]
```diff
@@ -0,0 +1,156 @@
+---
+title: 50 TiB 数据导入最佳实践
+summary: 本文根据导入大单表的经验，总结出了一套导入大量数据的最佳实践，希望对你在导入大数据量、大单表的场景有所帮助。
```
Comment (lilin90) [inline tidb-lightning/tidb-lightning-50TB-import-best-practices.md:3]:
```suggestion
summary: 根据导入大单表的经验总结出的导入大量数据的一套最佳实践。
```

File: tidb-lightning/tidb-lightning-50TB-import-best-practices.md [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+---
+
+# 50 TiB 数据导入最佳实践
+
+TiDB Lightning（[物理导入模式](/tidb-lightning/tidb-lightning-physical-import-mode.md)）是一款用于将离线数据导入空表、空集群的高效的数据导入工具。并且 TiDB Lightning 以文件作为数据源。TiDB Lightning 提供了两种运行方式：单实例和[并行导入](/tidb-lightning/tidb-lightning-distributed-import.md)，满足不同规模的源文件导入。
```
Comment (lilin90) [inline tidb-lightning/tidb-lightning-50TB-import-best-practices.md:8]:
```suggestion
TiDB Lightning（[物理导入模式](/tidb-lightning/tidb-lightning-physical-import-mode.md)）是一款用于将离线数据导入空表、空集群的高效的数据导入工具。TiDB Lightning 以文件作为数据源，提供了单实例和[并行导入](/tidb-lightning/tidb-lightning-distributed-import.md)两种运行方式，以满足不同规模的源文件导入。
```

File: tidb-lightning/tidb-lightning-50TB-import-best-practices.md [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+
+- 如果源文件数据规模在 10 TiB 以内，建议通过单个 TiDB Lightning 实例进行导入。
+- 如果源文件数据规模超过 10 TiB，建议通过多个 TiDB Lightning 实例进行[并行导入](/tidb-lightning/tidb-lightning-distributed-import.md)。
+- 如果源文件数据规模特别大（比如达到 50 TiB 及以上），在使用并行导入的同时，还需要针对源数据特点、表定义、参数配置等进行一定的准备和调优，才能更好、更快的完成大规模的数据导入。
```
Comment (lilin90) [inline tidb-lightning/tidb-lightning-50TB-import-best-practices.md:12]:
```suggestion
- 如果源文件数据规模特别大（比如达到 50 TiB 及以上），在使用并行导入的同时，还需要针对源数据特点、表定义、参数配置等进行一定的准备和调优，才能更好、更快地完成大规模的数据导入。
```

File: tidb-lightning/tidb-lightning-50TB-import-best-practices.md [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+
+本文主要介绍了影响 TiDB Lightning 数据导入的一些关键因素及操作步骤。我们在内部环境和客户现场都曾成功导入过 50 TiB 以上的大单表数据，基于这些真实的应用场景，沉淀了本文中的最佳实践。这些最佳实践可以帮助你成功导入大型数据。
```
Comment (lilin90) [inline tidb-lightning/tidb-lightning-50TB-import-best-practices.md:14]:
```suggestion
本文主要介绍影响 TiDB Lightning 数据导入的一些关键因素和操作步骤。PingCAP 在内部环境和客户现场都曾成功导入过 50 TiB 以上的大单表数据，基于这些真实的应用场景，沉淀了本文中的最佳实践，希望可以帮你成功导入大规模数据。
```

**Review by lilin90 — COMMENTED** [review]
I haven't finished my review and will continue soon later.

File: TOC.md [hunk 13]
```diff
@@ -103,6 +103,7 @@
 - 数据迁移
   - [数据迁移概述](/migration-overview.md)
   - [迁移工具](/migration-tools.md)
+  - [最佳实践](/tidb-lightning/tidb-lightning-50tb-data-import-best-practices.md)
```
[the remark this answers is not in the document]
Comment (lilin90) [reply TOC.md]: Resolved in https://github.com/pingcap/docs-cn/pull/14271/commits/9d47aa3c9667519a64e2bf9682aa600cd479b337.

File: tidb-lightning/data-import-best-practices.md [hunk 14]
```diff
@@ -0,0 +1,156 @@
+---
+title: 50 TiB 数据导入最佳实践
+summary: 根据导入大单表的经验总结出的导入大量数据的一套最佳实践。
```
Comment (lilin90) [inline tidb-lightning/data-import-best-practices.md:3]:
```suggestion
summary: 了解将大规模数据导入 TiDB 的最佳实践。
```

**Review by hfxsd — APPROVED** [review]

**Review by Frank945946 — APPROVED** [review]

File: tidb-lightning/data-import-best-practices.md [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+多表导入会导致 Checksum、ANALYZE 时间的增加，甚至超过数据导入本身，但是一般不需要调整配置。如果多表中存在单个或多个大表的情况，可以把这类大表的源文件划分出来，单独进行导入。
+
+本小节重点介绍大单表导入的最佳实践。大单表没有严格的定义，一般认为符合以下任一条件者即为大单表：
+
+- 大小超过 10 TiB
+- 行数超过 10 亿、列数超过 50 的宽表
+
+### 准备源文件
+
+根据上述源文件准备的步骤产生源文件，对于大单表，如果不能做到全局有序，但是可以做到文件内按主键有序，且是标准的 CSV 文件，可以尽量生成单个大文件（例如每个 20 GiB），然后开启 [`strict-format`](/migrate-from-csv-files-to-tidb.md#第-4-步导入性能优化可选)，既可以降低 TiDB Lightning 实例之间导入的数据文件中存在主键和唯一键的重叠，又能在导入前由 TiDB Lightning 实例对大文件进行切分，达到最佳的导入速度。
```
Comment (lilin90) [inline tidb-lightning/data-import-best-practices.md:128]:
```suggestion
根据上述[准备源文件的步骤](#准备源文件)产生源文件，对于大单表，如果不能做到全局有序，但是可以做到文件内按主键有序，且是标准的 CSV 文件，可以尽量生成单个大文件（例如每个 20 GiB），然后开启 [`strict-format`](/migrate-from-csv-files-to-tidb.md#第-4-步导入性能优化可选)，既可以降低 TiDB Lightning 实例之间导入的数据文件中存在主键和唯一键的重叠，又能在导入前由 TiDB Lightning 实例对大文件进行切分，达到最佳的导入速度。
```

File: tidb-lightning/data-import-best-practices.md [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+## 导入大单表的最佳实践
+
+多表导入会导致 Checksum、ANALYZE 时间的增加，甚至超过数据导入本身，但是一般不需要调整配置。如果多表中存在单个或多个大表的情况，可以把这类大表的源文件划分出来，单独进行导入。
+
+本小节重点介绍大单表导入的最佳实践。大单表没有严格的定义，一般认为符合以下任一条件者即为大单表：
+
+- 大小超过 10 TiB
+- 行数超过 10 亿、列数超过 50 的宽表
+
+### 准备源文件
```
Comment (lilin90) [inline tidb-lightning/data-import-best-practices.md:126]: 同一篇文档里尽量避免相同的文内标题，以免在引用到具体小节时造成困惑。
```suggestion
### 生成源文件
```

File: tidb-lightning/data-import-best-practices.md [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+    - [Issue-6426](https://github.com/tikv/pd/issues/6426)：PD [范围调度](/tidb-lightning/tidb-lightning-physical-import-mode-usage.md#导入时暂停-pd-调度的范围)接口存在未打散 Region 的情况，导致 Scatter Region 超时。v6.2.0 之前采用停止全局调度的方式，不会出现该问题。
+    - [Issue-43079](https://github.com/pingcap/tidb/pull/43079)：TiDB Lightning 对 NotLeader 错误的重试未刷新 Region Peers 信息。
+    - [Issue-43291](https://github.com/pingcap/tidb/issues/43291)：TiDB Lightning 未对临时文件删除的情况（"No such file or directory"）进行重试。
+
+## 准备源文件
+
+- 在生成文件时，在单个文件内尽量按照主键排序；如果表定义没有主键，可以添加一个自增主键，此时对文件内容顺序无要求。
+- 在给多个 TiDB Lightning 实例分配要导入的源文件时，尽量避免多个源文件之间存在重叠的主键或非空唯一索引的情况。如果生成文件是全局有序，可以按照范围划分不同的文件给不同 TiDB Lightning 实例进行导入，达到最佳导入效果。
+- 在生成文件时，每个文件尽量控制在 96 MiB 以下。
+- 如果文件特别大，超过 256 MiB，需要开启 [strict-format](/migrate-from-csv-files-to-tidb.md#第-4-步导入性能优化可选)。
```
Comment (lilin90) [inline tidb-lightning/data-import-best-practices.md:81]:
```suggestion
- 如果文件特别大，超过 256 MiB，需要开启 [`strict-format`](/migrate-from-csv-files-to-tidb.md#第-4-步导入性能优化可选)。
```

**Pull request merged**